Add torch.complex support in JIT

nikithamalgifb opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52929 Add torch.complex support in JIT**

Summary:
========
Adds Support for Torch.complex in JIT

Test:
====
python test/test_jit.py -k test_torch_complex

Commit:
Add torch.complex support in JIT

[ghstack-poisoned]

Commit:
Update on "Add torch.complex support in JIT"


Summary:
========
Adds Support for Torch.complex in JIT

Test:
====
python test/test_jit.py -k test_torch_complex

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1216,6 +1283,22 @@ RegisterOperators reg2({
         aliasAnalysisFromSchema()),
+
+#define DEFINE_COMPLEX_MIXED_OP(type_a, type_b)                      \
+  Operator(                                                          \
```
OOC, why not using the same macro to define all these different schemas?

nikithamalgifb replied:
So when type is `float` then , we need to declare the respective variable `real or img` here as `double` and read the stack variable into the `double` variable. Else if we just try to read the `float` variable directly from stack , then the output differs from eager mode. For example here:
 If we use the macro here `DEFINE_COMPLEX_MIXED_OP(float, float)` and try to call create a complex value. This is how it differs in eager and JIT.
```
complex(4.6, 7.2)

In eager: 4.6 + 7.2j

In JIT: 4.5555559 + 7.111123j
```

nikithamalgifb replied:
I don't think that `2` is possible. When we call complex( arg1, arg2) it tries to match the type of args with the `Operator definition`. If we end up passing one more `arg` to the macro, then that would mean complex.* would take 3 args instead of 2.

gmagogsfm commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1016,6 +1016,73 @@ RegisterOperators reg2({
+    Operator(
+        "aten::complex.float_bool(float real, bool img) -> complex",
+        [](Stack* stack) {
+          double real;
```
Please fix these clang-tidy warnings.

nikithamalgifb replied:
Agreed, Will do.

gmagogsfm commented on test/test_jit.py:
```diff
@@ -1085,6 +1085,172 @@ def fn(*args):
+        # The following operations are not allowed
+        # in Torchscript
+
+        with self.assertRaisesRegex(RuntimeError, "operation failed"):
```
What's the full error message like? is "operation failed" the most representative sub-string from that error?

nikithamalgifb replied:
Yes. "This operation failed in Torchscript"

gmagogsfm replied:
I tried to run this test case with your PR patched, here is the error message I got:

```
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "complex.py", line 5, in fn_tensor_float
@torch.jit.script
def fn_tensor_float(real, img: float):
    return complex(real, img)
           ~~~~~~~ <--- HERE
RuntimeError: Cannot input a tensor of dimension other than 0 as a scalar argument
```

It seems the last line is more representative. Let me know what you think.

nikithamalgifb replied:
I agree, the last line is more representative. But I wasn't sure if it would be a part of the TS runtime error. May be it is . I didn't try adding the last line, I'll give that a try and change.

gmagogsfm reviewed (commented):
Thanks for adding it. Generally looks good! Just have some minor cleanups needed.

Commit:
Update on "Add torch.complex support in JIT"


Summary:
========
Adds Support for Torch.complex in JIT

Test:
====
python test/test_jit.py -k test_torch_complex

[ghstack-poisoned]

Commit:
Update on "Add torch.complex support in JIT"


Summary:
========
Adds Support for Torch.complex in JIT

Test:
====
python test/test_jit.py -k test_torch_complex

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1216,6 +1286,22 @@ RegisterOperators reg2({
         hashValue,
         aliasAnalysisFromSchema()),
+
+#define DEFINE_COMPLEX_MIXED_OP(type_a, type_b)                      \
```
Thanks for fixing the clang-tidy issues, there seems to be two more warnings left, could you take a look? Thanks.

nikithamalgifb replied:
Yeah, this one suggests to define a 'constexpr' template function instead of a macro. Let me see if this can be disabled.

Commit:
Update on "Add torch.complex support in JIT"


Summary:
========
Adds Support for Torch.complex in JIT

Test:
====
python test/test_jit.py -k test_torch_complex

[ghstack-poisoned]

nikithamalgifb commented:
I somehow messed this PR . I am opening a new one.
